In [1]:
# Let's first do the necessary imports.

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

import os
print(os.listdir("../input"))
# Any results you write to the current directory are saved as output.

pd.options.mode.chained_assignment = None  # default='warn'

['d', 'titanic']


In [3]:
import seaborn as sns
import matplotlib.pyplot as plt
% matplotlib inline
import math

titanic_data = pd.read_csv("/kaggle/input/titanic/test.csv")

In [5]:
# Load the train and test datasets to create two dataframes
train = pd.read_csv('/kaggle/input/titanic/train.csv')
test = pd.read_csv('/kaggle/input/titanic/test.csv')

In [6]:
titanic_data.head(10)

PassengerId  Pclass   ...    Cabin Embarked
0          892       3   ...      NaN        Q
1          893       3   ...      NaN        S
2          894       2   ...      NaN        Q
3          895       3   ...      NaN        S
4          896       3   ...      NaN        S
5          897       3   ...      NaN        S
6          898       3   ...      NaN        Q
7          899       2   ...      NaN        S
8          900       3   ...      NaN        C
9          901       3   ...      NaN        S

[10 rows x 11 columns]

In [7]:
print("# of passengers in original data: " + str(len(titanic_data.index)))

# of passengers in original data: 418


**Time to understand the structure of your data**

.describe() summarizes the columns/features of the DataFrame, including the count of observations, mean, max and so on. 

.shape gives the dimensions of the dataframe

In [7]:
train.describe() 

PassengerId    Survived     ...           Parch        Fare
count   891.000000  891.000000     ...      891.000000  891.000000
mean    446.000000    0.383838     ...        0.381594   32.204208
std     257.353842    0.486592     ...        0.806057   49.693429
min       1.000000    0.000000     ...        0.000000    0.000000
25%     223.500000    0.000000     ...        0.000000    7.910400
50%     446.000000    0.000000     ...        0.000000   14.454200
75%     668.500000    1.000000     ...        0.000000   31.000000
max     891.000000    1.000000     ...        6.000000  512.329200

[8 rows x 7 columns]

In [8]:
train.shape

(891, 12)

**Data Dictionary** (to understand more please visit: https://www.kaggle.com/c/titanic/data)

      Variable         Definition                          Key

      survival         Survival                            0 = No, 1 = Yes
      pclass           Ticket class                        1 = 1st, 2 = 2nd, 3 = 3rd
      sex              Sex    
      Age              Age in years    
      sibsp            # of siblings / spouses aboard the Titanic    
      parch            # of parents / children aboard the Titanic    
      ticket           Ticket number    
      fare             Passenger fare    
      cabin            Cabin number    
      embarked         Port of Embarkation                 C = Cherbourg, Q = Queenstown, S=  Southampton



Before we dive deeper in, let's see if we can answer some basic questions from data directly. 

* **How many people in your training set survived the disaster with the Titanic? **
* **Number of males that survived vs number of males who did not survive**
* **Number of femailes that survived vs number of females who did not survive**
* **Does Age play a role?**

Analyzing Data

In [11]:
sns.countplot(x="Survived" , data=titanic_data)

ValueError: Could not interpret input 'Survived'

**How many people in your training set survived the disaster with the Titanic? **

In [9]:
# absolute numbers
print(train["Survived"].value_counts())

# percentages
print(train["Survived"].value_counts(normalize = True))

0    549
1    342
Name: Survived, dtype: int64
0    0.616162
1    0.383838
Name: Survived, dtype: float64


You will see that 549 individuals died which is about 61.61 % and 342 survived which is about 38.38 %.  

**Number of males that survived vs number of males who did not survive**

**Number of femailes that survived vs number of females who did not survive**

In [10]:
# Males that survived vs males that passed away
print(train["Survived"][train["Sex"] == 'male'].value_counts())

# Females that survived vs Females that passed away
print(train["Survived"][train["Sex"] == 'female'].value_counts())

# Normalized male survival
print(train["Survived"][train["Sex"] == 'male'].value_counts(normalize=True))

# Normalized female survival
print(train["Survived"][train["Sex"] == 'female'].value_counts(normalize=True))

0    468
1    109
Name: Survived, dtype: int64
1    233
0     81
Name: Survived, dtype: int64
0    0.811092
1    0.188908
Name: Survived, dtype: float64
1    0.742038
0    0.257962
Name: Survived, dtype: float64


It looks like it makes sense to predict that all females will survive, and all men will die. (A novice prediction!)

**Does age play a role?**

Another variable that could influence survival is age; since it's probable that children were saved first. You can test this by creating a new column with a categorical variable Child. Child will take the value 1 in cases where age is less than 18, and a value of 0 in cases where age is greater than or equal to 18.

In [11]:
# Create the column Child and assign to 'NaN'
train["Child"] = float('NaN')

# Assign 1 to passengers under 18, 0 to those 18 or older. Print the new column

train["Child"][train["Age"] < 18] = 1
train["Child"][train["Age"] >= 18] = 0

# Print normalized Survival Rates for passengers under 18
print(train["Survived"][train["Child"] == 1].value_counts(normalize = True))

# Print normalized Survival Rates for passengers 18 or older
print(train["Survived"][train["Child"] == 0].value_counts(normalize = True))

1    0.539823
0    0.460177
Name: Survived, dtype: float64
0    0.618968
1    0.381032
Name: Survived, dtype: float64


As you can see from the survival proportions, age does certainly seem to play a role.

**Simple Decision Tree**

We will need the following to build a decision tree
* target: A one-dimensional numpy array containing the target/response from the train data. (Survival in this case)
* features: A multidimensional numpy array containing the features/predictors from the train data. (ex. Sex, Age)


In [12]:
# Print the train data to see the available features
print(train.columns.values)

['PassengerId' 'Survived' 'Pclass' 'Name' 'Sex' 'Age' 'SibSp' 'Parch'
 'Ticket' 'Fare' 'Cabin' 'Embarked' 'Child']


Let us choose our target and feature variables. 
* Target: Survived Column
* Feature: Passenger, Class, Sex, Age, Fare
* Build a decision tree tree_one to predict survival using features and target.
* Look at the importance of features in your tree and compute the score 

Before we proceed further with Decision Tree we need to do some cleanup. Let's first check for missing values in any of the columns that we will be using in our feature array. 

In [13]:
# Import 'tree' from scikit-learn library
from sklearn import tree

Let's check for missing values. 

In [14]:
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
Child          177
dtype: int64

For our feature array we are using Age column and it has 177 missing values. Let's first impute values on this column with the median of all present values.

In [15]:
train["Age"] = train["Age"].fillna(train["Age"].median())

In [16]:
# Let us take another look at the values in dataset again.
train.head()

PassengerId  Survived  Pclass  ...   Cabin Embarked  Child
0            1         0       3  ...     NaN        S    0.0
1            2         1       1  ...     C85        C    0.0
2            3         1       3  ...     NaN        S    0.0
3            4         1       1  ...    C123        S    0.0
4            5         0       3  ...     NaN        S    0.0

[5 rows x 13 columns]

Sex and Embarked are categorical values but they are in non-numeric format. We need to convert them to numeric format. Emarked also has some missing values and we need to impute them with the majority value that is 'S' before converting to numeric format. 

In [17]:
# Checking the frequency of values in Embarked column. 
# Thank you @A.mannan.z for mentioning that this step was not added. 
train['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [18]:
# Convert the male and female groups to integer form
train["Sex"][train["Sex"] == "male"] = 0
train["Sex"][train["Sex"] == "female"] = 1

# Impute the Embarked variable
train["Embarked"] = train["Embarked"].fillna("S")

# Convert the Embarked classes to integer form
train["Embarked"][train["Embarked"] == "S"] = 0
train["Embarked"][train["Embarked"] == "C"] = 1
train["Embarked"][train["Embarked"] == "Q"] = 2

In [19]:
# Let's take a look at train.head() again. 
train.head()

PassengerId  Survived  Pclass  ...   Cabin Embarked  Child
0            1         0       3  ...     NaN        0    0.0
1            2         1       1  ...     C85        1    0.0
2            3         1       3  ...     NaN        0    0.0
3            4         1       1  ...    C123        0    0.0
4            5         0       3  ...     NaN        0    0.0

[5 rows x 13 columns]

It looks good, so let's proceed with creating target and features array.

In [20]:
# Create the target and features numpy arrays: target, features_one
target = train["Survived"].values
features_one = train[["Pclass", "Sex", "Age", "Fare"]].values

# Fit your first decision tree: my_tree_one
my_tree_one = tree.DecisionTreeClassifier()
my_tree_one = my_tree_one.fit(features_one, target)

# Look at the importance and score of the included features
print(my_tree_one.feature_importances_)
print(my_tree_one.score(features_one, target))

[0.12566688 0.31274009 0.23458607 0.32700696]
0.9775533108866442


Based on our tree, we can see that "Fare" contributed more than other features. That is an analysis we are driving out from this decision tree. 

Let's now take a look at our test dataset and see if there are any null values. 

In [21]:
# Looking at test.info() we see that Fare has one null value. 
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Age has 86 null values and Fare has 1 null value. Let's do imputation similar to what we did for train set. 

In [22]:
test["Fare"] = test["Fare"].fillna(test["Fare"].median())
test["Age"] = test["Age"].fillna(test["Age"].median())

test["Sex"][test["Sex"] == "male"] = 0
test["Sex"][test["Sex"] == "female"] = 1

Let's extract features from test set and make our prediction based on my_tree_one model.

In [23]:
# Extract the features from the test set: Pclass, Sex, Age, and Fare.
test_features = test[['Pclass', 'Sex', 'Age', 'Fare']].values

# Make your prediction using the test set
my_prediction = my_tree_one.predict(test_features)
print(my_prediction)

# Create a data frame with two columns: PassengerId & Survived. Survived contains your predictions
PassengerId =np.array(test["PassengerId"]).astype(int)
my_solution = pd.DataFrame(my_prediction, PassengerId, columns = ["Survived"])
print(my_solution)

# Check that your data frame has 418 entries
print(my_solution.shape)

# Write your solution to a csv file with the name my_solution.csv
my_solution.to_csv("my_solution_one.csv", index_label = ["PassengerId"])

[0 0 1 1 1 0 0 0 1 0 0 0 1 1 1 1 0 1 1 0 0 1 1 0 1 0 1 1 1 0 0 0 1 0 1 0 0
 0 0 1 0 1 0 1 1 0 0 0 1 1 1 0 1 1 1 0 0 0 1 1 0 0 0 1 1 0 1 0 0 1 1 0 1 0
 1 0 0 1 0 1 1 0 0 0 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 0 1 0 0 0 1 0 0 0 1 0 0
 0 1 1 1 0 1 1 0 1 1 0 1 0 0 1 0 1 0 0 1 0 0 1 0 0 1 0 0 0 0 0 0 0 0 1 1 0
 1 0 1 0 0 1 0 0 1 1 0 1 1 1 1 1 0 1 1 0 0 0 0 1 0 1 0 1 1 0 1 1 0 1 1 0 1
 0 1 0 0 0 0 0 1 0 1 0 1 0 0 0 0 1 0 1 0 0 0 0 1 0 1 1 0 0 0 0 1 0 1 0 1 0
 1 1 1 0 0 1 0 0 0 1 0 0 0 0 0 1 1 1 1 1 1 0 0 0 1 0 1 0 1 0 0 0 0 0 0 0 1
 0 0 0 1 1 0 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 1 1 0 1 0 0 0 1 0 1 0 1 0 0 0
 1 0 0 0 0 0 0 0 1 1 0 1 1 0 0 1 0 0 1 1 0 0 0 0 0 0 0 1 1 0 1 0 0 0 1 0 1
 1 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 1 1 0 0 0 1 0 1 0 0 1 0 1 1 1 1 0 0 0 1 0
 0 1 0 0 1 1 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 1 1 0 0 1 0 1 0 0 1 0 1 1 0 0 0
 0 1 1 1 1 0 0 1 0 0 0]
      Survived
892          0
893          0
894          1
895          1
896          1
897          0
898          0
899          0
900          1
9

In [24]:
!ls

my_solution_one.csv
